In [5]:
!pip uninstall -y torch torchvision torchaudio torch-scatter torch-sparse torch-cluster torch-spline-conv torch-geometric



Found existing installation: torch 2.5.0+cu118
Uninstalling torch-2.5.0+cu118:
  Successfully uninstalled torch-2.5.0+cu118
Found existing installation: torchvision 0.20.0+cu118
Uninstalling torchvision-0.20.0+cu118:
  Successfully uninstalled torchvision-0.20.0+cu118
Found existing installation: torchaudio 2.5.0+cu118
Uninstalling torchaudio-2.5.0+cu118:
  Successfully uninstalled torchaudio-2.5.0+cu118
Found existing installation: torch-geometric 2.6.1
Uninstalling torch-geometric-2.6.1:
  Successfully uninstalled torch-geometric-2.6.1


In [6]:
# pip install torch==2.1.0 torchvision torchaudio --index-url https://download.pytorch.org/whl/cpu  # For CPU
!pip install torch==2.1.0 torchvision torchaudio --index-url https://download.pytorch.org/whl/cu118  # For GPU



Looking in indexes: https://download.pytorch.org/whl/cu118
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.3/2.3 GB 9.7 MB/s eta 0:00:000m eta 0:00:010:00:06
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 89.3/89.3 MB 10.8 MB/s eta 0:00:00m eta 0:00:010:01:02
INFO: pip is looking at multiple versions of torchvision to determine which version is compatible with other requirements. This could take a while.
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.5/6.5 MB 9.8 MB/s eta 0:00:0010.0 MB/s eta 0:00:01
  Using cached https://download.pytorch.org/whl/cu118/torchvision-0.20.0%2Bcu118-cp39-cp39-linux_x86_64.whl (6.5 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.3/6.3 MB 5.6 MB/s eta 0:00:00 MB/s eta 0:00:01:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.3/6.3 MB 9.9 MB/s eta 0:00:001 MB/s eta 0:00:01:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.3/6.3 MB 10.9 MB/s eta 0:00:00 MB/s eta 0:00:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.3/6.3 MB 11.4 MB/s eta 0:00:00 MB/s 

In [7]:
# !pip install torch-scatter torch-sparse torch-cluster torch-spline-conv torch-geometric -f https://data.pyg.org/whl/torch-2.1.0+cpu.html
!pip install torch-scatter torch-sparse torch-cluster torch-spline-conv torch-geometric -f https://data.pyg.org/whl/torch-2.1.0+cu118.html


Looking in links: https://data.pyg.org/whl/torch-2.1.0+cu118.html
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.2/10.2 MB 6.2 MB/s eta 0:00:00MB/s eta 0:00:01:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.9/4.9 MB 4.3 MB/s eta 0:00:00 MB/s eta 0:00:01:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.3/3.3 MB 3.6 MB/s eta 0:00:00 MB/s eta 0:00:01:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 888.3/888.3 kB 2.2 MB/s eta 0:00:002.0 MB/s eta 0:00:01
  Using cached torch_geometric-2.6.1-py3-none-any.whl.metadata (63 kB)
Using cached torch_geometric-2.6.1-py3-none-any.whl (1.1 MB)


In [13]:
!pip install seaborn

In [8]:
import torch
print(torch.__version__)


2.1.0+cu118


In [9]:
import torch
import torch_geometric
import torch_cluster

print("Torch version:", torch.__version__)
print("Torch Geometric version:", torch_geometric.__version__)
print("CUDA Available:", torch.cuda.is_available())
print("Torch Cluster is installed correctly!")


Torch version: 2.1.0+cu118
Torch Geometric version: 2.6.1
CUDA Available: True
Torch Cluster is installed correctly!


In [10]:
!pip install numpy==1.26.4


In [11]:
import numpy as np
print("NumPy version:", np.__version__)


NumPy version: 1.26.4


In [14]:
import numpy as np
import torch
import seaborn as sns
import matplotlib.pyplot as plt
import pandas as pd
from sklearn.model_selection import train_test_split

import torch.nn.functional as F
from torch_geometric.nn import DynamicEdgeConv, GATConv, global_max_pool
from torch_geometric.data import Data, DataLoader

In [16]:
# Load dataset
file_path = ("/home/ujjwal/Dashboard/GAN on HEP/GNN on HEP/QG_jets.npz")
data = np.load(file_path, allow_pickle=True)

print(data.files)
# train_data = data['training_input']
# test_data = data['test_input']


['X', 'y']


In [17]:
data['X'].shape, data['y'].shape

((100000, 139, 4), (100000,))

In [18]:
# Extract training and test data
X = data['X']
y = data['y']

In [19]:
X[0].shape

(139, 4)

In [20]:
y[:-5]

array([1., 1., 1., ..., 1., 0., 0.])

In [21]:
_, X_new, _, y_new = train_test_split(X, y, test_size=0.2, random_state=42)

In [22]:
X_train, X_test, y_train, y_test = train_test_split(X_new, y_new, test_size=0.2, random_state=42)

In [23]:
X_train.shape, X_test.shape, y_train.shape, y_test.shape

((16000, 139, 4), (4000, 139, 4), (16000,), (4000,))

In [24]:
X_train[0]

array([[ 8.19062772e-02,  7.13226663e-01,  2.65543418e+00,
         2.20000000e+01],
       [ 1.49909908e-01,  9.02892302e-01,  2.88856597e+00,
         2.20000000e+01],
       [ 6.13477869e-01,  8.47353599e-01,  2.84087708e+00,
         1.30000000e+02],
       [ 1.05701503e+00,  5.76321842e-01,  2.80262170e+00,
         2.20000000e+01],
       [ 1.86652044e+00,  7.51513249e-01,  2.98552541e+00,
         2.20000000e+01],
       [ 5.90934964e-01,  4.73384081e-01,  2.89736364e+00,
         2.20000000e+01],
       [ 3.90991255e+00,  5.26869874e-01,  3.01202014e+00,
         2.11000000e+02],
       [ 2.11268998e+01,  5.92743577e-01,  3.04135006e+00,
        -2.11000000e+02],
       [ 4.01586158e+00,  5.41922258e-01,  2.95194321e+00,
         2.20000000e+01],
       [ 9.60884753e+00,  6.37358826e-01,  3.02288679e+00,
         2.11000000e+02],
       [ 8.50202136e+00,  6.12128632e-01,  3.02934779e+00,
        -2.11000000e+02],
       [ 1.29459461e+01,  6.25134101e-01,  2.92785380e+00,
      

### Models

The considerations we take when converting the point-cloud dataset into a graph structure, where individual points become nodes and relationships between them become edges.

When projecting each point-cloud sample into a graph, every point is treated as a node. In this dataset, each sample contains 139 points with 4 features each (for example, spatial coordinates and other physical properties

). This configuration ensures that the graph’s node features faithfully capture all the measured attributes of the original point-cloud.

A principal design decision is how to define the connectivity between nodes. One effective approach is to compute all pairwise connections (using a method like torch.combinations) to generate a fully connected graph. While this means that every node is theoretically linked to every other node, models such as DynamicEdgeConv can then dynamically select and focus on the k-nearest neighbors (with a preset value like k = 8) during message passing. This strategy preserves comprehensive interaction possibilities while allowing the network to learn localized patterns without the computational burden of a naive full connectivity during every operation

.

Other key considerations include:

    Uniformity and Consistency:

    Since all samples have the same number of points (139), a single precomputed structure for the edge list can be used across the dataset. This ensures that each graph has consistent connectivity, which is beneficial for batching and for the downstream graph neural network layers.

    Feature Preservation:

    The node features are kept intact without reduction or masking at this stage, ensuring that every attribute—from spatial coordinates to other numerical measurements—is available for the model to learn meaningful patterns. This retention is crucial when dealing with complex physical data represented in the point-cloud.

    Computational Efficiency:

    Although computing a fully connected graph can be computationally heavy in very large point clouds, the moderate size of 139 nodes keeps the process tractable. Moreover, using dynamic edge selection methods within the network helps streamline computations by focusing on the most relevant neighboring relationships.

    Handling Padded or Invalid Data:

    The dataset sample provided includes rows that are entirely zeros. These entries likely represent padding or missing values. One must decide whether to filter these out or design the model to ignore them, ensuring that they do not adversely affect the learning of the true inter-point relationships.

    Integration with Graph Neural Networks:
    
    Finally, the conversion process is designed with frameworks like PyTorch Geometric in mind, where each graph object includes not only the node features and edge indices but also a batch attribute. This attribute is critical for proper handling during training when multiple graphs are processed together.


In [25]:
def preprocess_data(X, y):
    data_list = []
    # Precompute the edge_index for a fully connected graph on 139 nodes
    nodes = torch.arange(X.shape[1])
    edge_index = torch.combinations(nodes, r=2).t()  # shape: [2, num_edges]
    for i in range(X.shape[0]):
        # Create node features tensor
        x = torch.tensor(X[i], dtype=torch.float)
        y_label = torch.tensor([y[i]], dtype=torch.long)
        data_list.append(Data(x=x, edge_index=edge_index, y=y_label))
    return data_list

In [26]:
class DGCNN(torch.nn.Module):
    def __init__(self, k=8, input_dim=4, hidden_dim=64, output_dim=2):
        super(DGCNN, self).__init__()
        self.conv1 = DynamicEdgeConv(
            torch.nn.Sequential(
                torch.nn.Linear(2 * input_dim, hidden_dim),
                torch.nn.ReLU(),
                torch.nn.Linear(hidden_dim, hidden_dim)
            ),
            k=k
        )
        self.conv2 = DynamicEdgeConv(
            torch.nn.Sequential(
                torch.nn.Linear(2 * hidden_dim, hidden_dim),
                torch.nn.ReLU(),
                torch.nn.Linear(hidden_dim, hidden_dim)
            ),
            k=k
        )
        self.fc = torch.nn.Linear(hidden_dim, output_dim)

    def forward(self, x, batch):
        x = self.conv1(x, batch)
        x = self.conv2(x, batch)
        # Use global pooling to aggregate node-level features per graph
        x = global_max_pool(x, batch)
        return self.fc(x)

Here, for GAT model we have reduced the hidden dim from 64 to 32 and heads from 8 to 4 due to limited GPU power. And for this we have to reduce the batch_size also to 2 from 32

In [53]:
class GAT(torch.nn.Module):
    def __init__(self, input_dim=4, hidden_dim=32, output_dim=2, heads=4):
        super(GAT, self).__init__()
        self.conv1 = GATConv(input_dim, hidden_dim, heads=heads, concat=True)
        self.conv2 = GATConv(hidden_dim * heads, hidden_dim, heads=heads, concat=False)
        self.fc = torch.nn.Linear(hidden_dim, output_dim)

    def forward(self, x, edge_index, batch):
        x = self.conv1(x, edge_index)
        x = F.elu(x)
        x = self.conv2(x, edge_index)
        # Aggregate node features to get a graph-level embedding
        x = global_max_pool(x, batch)
        return self.fc(x)

In [54]:
def train(model, loader, optimizer, criterion, device):
    model.train()
    total_loss = 0
    for data in loader:
        data = data.to(device)
        optimizer.zero_grad()
        # For GAT we pass (x, edge_index, batch), while for DGCNN we pass (x, batch)
        if isinstance(model, GAT):
            out = model(data.x, data.edge_index, data.batch)
        else:
            out = model(data.x, data.batch)
        loss = criterion(out, data.y)
        loss.backward()
        optimizer.step()
        total_loss += loss.item()
    return total_loss / len(loader)


In [55]:
# Evaluation Function
def evaluate(model, loader, device):
    model.eval()
    correct = 0
    total = 0
    with torch.no_grad():
        for data in loader:
            data = data.to(device)
            # out = model(data.x, data.edge_index if isinstance(model, GAT) else data.batch)
            
            if isinstance(model, GAT):
                out = model(data.x, data.edge_index, data.batch)  # Include batch
            else:
                out = model(data.x, data.batch)
            pred = out.argmax(dim=1)
            correct += (pred == data.y).sum().item()
            total += data.y.size(0)
    return correct / total


### Execution

In [30]:
# Load and preprocess data
train_data = preprocess_data(X_train, y_train)
test_data = preprocess_data(X_test, y_test)

/tmp/ipykernel_9673/697128300.py:9: DeprecationWarning: an integer is required (got type numpy.float64).  Implicit conversion to integers using __int__ is deprecated, and may be removed in a future version of Python.
  y_label = torch.tensor([y[i]], dtype=torch.long)


In [31]:
# Create DataLoader
batch_size = 32
train_loader = DataLoader(train_data, batch_size=batch_size, shuffle=True)
test_loader = DataLoader(test_data, batch_size=batch_size, shuffle=False)


/home/ujjwal/.conda/envs/myenv/lib/python3.9/site-packages/torch_geometric/deprecation.py:26: UserWarning: 'data.DataLoader' is deprecated, use 'loader.DataLoader' instead
  warnings.warn(out)


Evaluation of DCGNN

In [34]:
# Define device
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

# initialize model
model = DGCNN().to(device)

# Define optimizer and loss function
optimizer = torch.optim.Adam(model.parameters(), lr=0.0001)
criterion = torch.nn.CrossEntropyLoss()

# Train model
num_epochs = 10
for epoch in range(num_epochs):
    loss = train(model, train_loader, optimizer, criterion, device)
    print(f"Epoch {epoch+1}, Loss: {loss:.4f}")

# Evaluate model
accuracy = evaluate(model, test_loader, device)
print(f"Test Accuracy: {accuracy:.4f}")

Epoch 1, Loss: 0.7746
Epoch 2, Loss: 0.5409
Epoch 3, Loss: 0.5186
Epoch 4, Loss: 0.5254
Epoch 5, Loss: 0.5164
Epoch 6, Loss: 0.5019
Epoch 7, Loss: 0.5028
Epoch 8, Loss: 0.5033
Epoch 9, Loss: 0.5003
Epoch 10, Loss: 0.4918
Test Accuracy: 0.7638


Evaluation of GAT

In [56]:
import torch
torch.cuda.empty_cache()


In [57]:
# Create DataLoader
batch_size = 2
train_loader = DataLoader(train_data, batch_size=batch_size, shuffle=True)
test_loader = DataLoader(test_data, batch_size=batch_size, shuffle=False)

In [62]:
# Define device
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

# initialize model
model = GAT().to(device)
# Define optimizer and loss function
optimizer = torch.optim.Adam(model.parameters(), lr=0.0001)
criterion = torch.nn.CrossEntropyLoss()

# Train model
num_epochs = 10
for epoch in range(num_epochs):
    loss = train(model, train_loader, optimizer, criterion, device)
    print(f"Epoch {epoch+1}, Loss: {loss:.4f}")

# # Evaluate model
# accuracy = evaluate(model, test_loader, device)
# print(f"Test Accuracy: {accuracy:.4f}")

Epoch 1, Loss: 0.9775
Epoch 2, Loss: 0.8697
Epoch 3, Loss: 0.7862
Epoch 4, Loss: 0.7407
Epoch 5, Loss: 0.6981
Epoch 6, Loss: 0.6526
Epoch 7, Loss: 0.6412
Epoch 8, Loss: 0.6021
Epoch 9, Loss: 0.5917
Epoch 10, Loss: 0.5886


In [63]:
model.to("cpu")
accuracy = evaluate(model, test_loader, "cpu")
model.to(device)  # Move back to GPU for training


GAT(
  (conv1): GATConv(4, 32, heads=4)
  (conv2): GATConv(128, 32, heads=4)
  (fc): Linear(in_features=32, out_features=2, bias=True)
)

In [64]:
print(f"Test Accuracy: {accuracy:.4f}")

Test Accuracy: 0.6877


# Comparative Analysis of DGCNN vs. GAT for Jet Particle Classification

## Performance Overview
| Metric          | DGCNN            | GAT              |
|-----------------|------------------|------------------|
| Test Accuracy   | **76.38%**       | **68.77%**       |
| Training Epochs | 10               | 10               |
| Final Loss      | 0.4918           | 0.5886           |

---

## **DGCNN Performance (76.38%)**
### Strengths
- **Local Feature Learning**  
  k-NN graph construction (k=8) effectively captures localized spatial patterns in particle jets.
- **Robust Convergence**  
  Stable loss reduction from 0.7746 → 0.4918 shows reliable optimization.
- **Computational Efficiency**  
  Handled full configuration (batch_size=32, hidden_dim=64) without memory constraints.

### Limitations
- **Fixed k Restriction**  
  Predefined neighborhood size may miss variable-scale particle interactions.

---

## **GAT Performance (68.77%)**
### Strengths
- **Theoretical Potential**  
  Attention mechanisms could capture non-local particle relationships.
- **Parameter Efficiency**  
  Achieved moderate accuracy despite severe constraints:


batch_size = 2 # Original: 32

hidden_dim = 32 # Original: 64

heads = 4 # Original: 8


### Limitations
- **Hardware Limitations**  
75% reduction in attention heads critically constrained model capacity.
- **Slow Convergence**  
Loss decreased gradually (0.9775 → 0.5886), needing more epochs.

---

## Key Comparative Insights
| Factor               | DGCNN Advantage                          | GAT Challenge                          |
|----------------------|------------------------------------------|----------------------------------------|
| **Data Efficiency**  | Faster initial convergence               | Requires 2-3× more epochs              |
| **Hardware Demands** | Efficient at batch_size=32               | Accuracy-limited at batch_size=2       |
| **Physics Alignment**| Distance-based edges match jet topology  | Attention needs domain adaptation      |

---

## Recommendations
### For DGCNN
1. Implement adaptive k-values using learned radius queries
2. Test hierarchical graph constructions

### For GAT


1. Add edge pruning for zero-padded nodes
2. Use masked attention for invalid particles

### Architectural Innovation
- Hybrid approach: DGCNN layers → GAT layers
- Graph U-Net with attention skip connections

---

## Conclusion
**DGCNN** currently demonstrates superior performance (76.38% vs 68.77%) under these constraints, but **GAT** shows untapped potential with proper resource allocation. Both models would benefit from explicit handling of zero-padded nodes observed in the dataset samples. Future work should focus on adaptive neighborhood selection for DGCNN and attention pattern specialization for GAT.
